File: advent_of_code/2022/day_04/solution.py
from elf import Elf
from utils.py_utils.constants import INPUT
from utils.py_utils.decorators import strip_newlines, split_lines_by, to_list


@to_list
@split_lines_by("[,-]")
@strip_newlines
def get_data(variation):
    with open(f"inputs/04_{variation}.txt") as f:
        data = f.readlines()
    return data


def modify(data):
    elves = []
    for first_start, first_end, second_start, second_end in data:
        first_elf = Elf(int(first_start), int(first_end))
        second_elf = Elf(int(second_start), int(second_end))
        elves.append([first_elf, second_elf])

    return elves


def easy(data):
    count = 0
    elves = modify(data)
    for first_elf, second_elf in elves:
        if first_elf.contains(second_elf) or second_elf.contains(first_elf):
            count += 1

    return count


def hard(data):
    count = 0
    elves = modify(data)
    for first_elf, second_elf in elves:
        if first_elf.overlaps(second_elf) or second_elf.overlaps(first_elf):
            count += 1

    return count


if __name__ == "__main__":
    data = get_data(INPUT)
    print(easy(data))
    print(hard(data))

File: advent_of_code/2022/day_04/solution.rb
require_relative "elf"
require_relative "../utils/ruby_utils/constants"


def raw_data(variation)
  File.readlines("inputs/04_#{variation}.txt",  chomp: true)
end

def modify(data)
  data.map do |line|
    /(\d+)-(\d+),(\d+)-(\d+)/ =~ line
    [Elf.new($1.to_i, $2.to_i), Elf.new($3.to_i, $4.to_i)]
  end
end

def get_data(variation)
  modify(raw_data(variation))
end

def easy(data)
  count = 0
  data.each do |first_elf, second_elf|
    if first_elf.contains?(second_elf) or second_elf.contains?(first_elf)
      count += 1
    end
  end

  count
end

def hard(data)
  count = 0
  data.each do |first_elf, second_elf|
    if first_elf.overlaps?(second_elf) or second_elf.overlaps?(first_elf)
      count += 1
    end
  end

  count
end

if __FILE__ == $0
  data = get_data(INPUT)
  puts easy(data)
  puts hard(data)
end
